src/rust/core/server/src/threads/worker/storage.rs

```diff
@@ -88,67 +89,86 @@ where
         loop {
             STORAGE_EVENT_LOOP.increment();
 
+            let now = Instant::now();
             self.storage.expire();
-
-            // get events with timeout
-            if self.poll.poll(&mut events, timeout).is_err() {
-                error!("Error polling");
+            if now.elapsed() > rustcommon_time::Duration::from_micros(500) {
+                debug!("expiration: {:?}", now.elapsed());
             }
 
-            if !events.is_empty() {
-                let mut worker_pending = self.worker_queues.pending();
-
-                trace!("handling events");
-                let mut empty = false;
-
-                while !empty {
-                    empty = true;
-                    for id in 0..workers {
-                        if worker_pending[id] > 0 {
-                            if let Ok(message) = self.worker_queues.recv_from(id) {
-                                trace!("handling request from worker: {}", id);
-                                PROCESS_REQ.increment();
-                                let token = message.token();
-                                let response = self.storage.execute(message.into_inner());
-                                let mut message = TokenWrapper::new(response, token);
-                                for retry in 0..QUEUE_RETRIES {
-                                    if let Err(QueueError::Full(m)) =
-                                        self.worker_queues.send_to(id, message)
-                                    {
-                                        if (retry + 1) == QUEUE_RETRIES {
-                                            error!("error sending message to worker");
-                                        }
-                                        let _ = self.worker_queues.wake(id);
-                                        message = m;
-                                    } else {
-                                        break;
+            // get events with timeout
+            // let now = Instant::now();
+            // if self.poll.poll(&mut events, timeout).is_err() {
+            //     error!("Error polling");
+            // }
+            // debug!("poll: {:?}", now.elapsed());
+
+            // if !events.is_empty() {
+            let mut worker_pending = self.worker_queues.pending();
+
+            trace!("handling events");
+            let mut empty = false;
+
+            while !empty {
+                empty = true;
+                for id in 0..workers {
+                    if worker_pending[id] > 0 {
+                        if let Ok(message) = self.worker_queues.recv_from(id) {
```

Comment:
If we start breaking down the processing time at different stages and/or on different threads, we should also consider the "outer" duration vs. "inner" duration. The "outer" duration could be the time it takes from writing a request to the (storage) queue to finishing reading the corresponding response; the "inner" duration is what you are measuring here. This will give us a sense of processing latency versus queuing latency. The former indicates a slow step on the storage thread, the latter indicates when we need to consider scale out the storage thread (through partitioning, etc). 

This will probably require a change in the type of content passed through the queue, instead of a simple `Request` object, we will need a `(Request, timestamp)` tuple.